### PR TITLE
Ankaios initial contribution with running scenario

### PR DESCRIPTION
**Pull request by inf17101**
Issue: #10

### CODE REVIEW

**Comment by inf17101** [comment]
@ladatz: Within this PR you can make any further changes required for the in-vehicle-stack and the additional applications for the scenarios.

**Comment by ladatz** [comment]
> @ladatz: Within this PR you can make any further changes required for the in-vehicle-stack and the additional applications for the scenarios.

Thanks @inf17101 , I will work on getting our images incorporated with this branch

**Comment by ladatz** [comment]
> > @ladatz: Within this PR you can make any further changes required for the in-vehicle-stack and the additional applications for the scenarios.
> 
> Thanks @inf17101 , I will work on getting the images incorporated with this branch

The images are currently located here: https://github.com/ladatz?tab=packages (I don't believe I have write permissions to add them to the blueprint container registry). My fork also includes the correct configuration for ankaios: https://github.com/ladatz/software-orchestration/tree/initial_ankaios_changes. I could not push the changes here; I am seeing 403. I removed the cloud connector from the startupState for this initial contribution. We can consider adding it back in in a later phase.

**Review by ladatz — APPROVED** [review]

**Comment by inf17101** [comment]
@ladatz: Thank you for contributing the changes.
I try this out tomorrow, too. Afterwards I will approve.

**Comment by inf17101** [comment]
The configuration changes are working. LGTM 👍 

@eriksven: Could you review the PR and merge it ?

File: eclipse-ankaios/README.md
Comment (eriksven) [inline eclipse-ankaios/README.md:82]: I starting to get confused by the actual name of this blueprint which I assumed would be either software-orchestration or maestro. Here you seem to reference with ankaio_blueprint or simply "blueprint". To my understanding this software orchestration blueprint is not exclusive to ankaios and there are other blueprints so it would be great to avoid confusion for people working on multiple blueprints too.

File: eclipse-ankaios/config/startupState.yaml
Comment (eriksven) [inline eclipse-ankaios/config/startupState.yaml:9]: To me it seems that they're a more favorable approaches instead of pulling from a personal GitHub container registry. Especially, since the pulled content/code is already from an Eclipse open source project. Would it be an alternative to directly pull from the Eclipse Chariott repository instead? @ladatz

**Comment by eriksven** [comment]
I added two comments and would wait for your thoughts on them before merging.

File: eclipse-ankaios/config/startupState.yaml
Comment (ladatz) [reply eclipse-ankaios/config/startupState.yaml:9]: This is a good point, however we are a little blocked by our permissions. We do not want to keep these images in my personal GitHub container registry long-term. We also do not want to host them in the individual projects' Container Registries for now either, because they have blueprint-specific settings and configurations to integrate with the other projects. Using the personal registry was an intermediate step to unblock our testing and development. 

We would like to host them in the eclipse-sdv-blueprints container registry (like what FMS does), but since we are not committers (and do not have write access), I get permission_denied when I try to push Images there.

I can see a few options on how to proceed: 

1. We can have someone with write access to the container registry pull down those images and push them into the blueprints container registry and then update these references to point to those images.
2. We can use this personal GitHub container registry for testing and development for now, merge this PR, and then once we can become committers to the project I can push them to the blueprint container registry myself.

I am open to any other discussion/ suggestions

**Comment by inf17101** [comment]
> @inf17101 , I added two comments and would wait for your thoughts on them before merging.

Yes thank you. Regarding the private images, I wanted to write you anyway. In our project we push all images into the organization's package repository, which is in my opinion the best way. However, it seems like that @ladatz does not have access to push an image. Could you give us access or enable it via self-service?

File: eclipse-ankaios/README.md
Comment (inf17101) [reply eclipse-ankaios/README.md:82]: We can rename it. The name should be kept general as there are different scenarios later. Then I suggest to mount it under "/workspaces/software-orchestration".

File: eclipse-ankaios/config/startupState.yaml
Comment (ladatz) [reply eclipse-ankaios/config/startupState.yaml:9]: A couple more comments: these are public images in a public container registry, but it is listed under my personal organization. I chose this container registry because it is most similar to the way that we have to do PRs without committer permissions: we have to create a fork in our personal organization.

Comment (inf17101) [reply eclipse-ankaios/config/startupState.yaml:9]: I would prefer that we become write access by becoming official committers. Then you can tag the images again with something like `docker image tag <image_id> ghcr.io/eclipse-sdv-blueprints/software-orchestration/eclipse-chariott/service-discovery:0.2.0`and push it to the package repository. All the temporary access would work of course somehow, but we shall also add some automatic release build later which is basically a container image published to the package repository as well.

Comment (eriksven) [reply eclipse-ankaios/config/startupState.yaml:9]: Ok, I see where this is coming from.

In general I would prefer it if the build process for the containers is fully transparent and it is comprehensible for an outsider what actually has been put into the containers.
Because of that my proposal is that you push the Dockerfile including the configuration for the build to the blueprint repository. You can then create a GitHub action for building and then pushing the container to the GHCR of this blueprint. This way no developer needs to locally build or manually push the container somewhere and it is clear what has been built.

This is also the approach we took for the containers in the FMS-Blueprint: https://github.com/eclipse-sdv-blueprints/fleet-management/blob/main/.github/workflows/containerize_fms_forwarder.yaml . So you could use these actions as a template.

I cannot give you specific rights to the repository. The rights are managed by the Eclipse Foundation and the only way to get some kind of writing right to the repository is by being a commiter.

Comment (ladatz) [reply eclipse-ankaios/config/startupState.yaml:9]: Thanks for the suggestion. I will look at our options (since our use case is slightly different than the FMS case, the code is hosted in other repos) to try to unblock ourselves.

Comment (ladatz) [reply eclipse-ankaios/config/startupState.yaml:9]: I am still blocked by permissions when trying to run the worflow that I am working on from my fork (the only way I can test at the moment). The workflow worked to push images to "my" ladatz GitHub container registry, but not to the eclipse-sdv-blueprints registry. I still get a 403.

Comment (eriksven) [reply eclipse-ankaios/config/startupState.yaml:9]: My guess is that it is intended to only allow pushes from within the context of the eclipse-sdv-blueprints repository and registry. My hope would be that the same workflow works when you contribute and run it in this repository. Would you be willing to contribute the workflow?

Comment (ladatz) [reply eclipse-ankaios/config/startupState.yaml:9]: Thanks @eriksven PR here: https://github.com/eclipse-sdv-blueprints/software-orchestration/pull/15

**Comment by eriksven** [comment]
I merged #15 and triggered the workflow for building the image. It is now available in https://github.com/eclipse-sdv-blueprints/software-orchestration/pkgs/container/software-orchestration%2Feclipse-chariott%2Fservice-discovery . 
Could adopt this PR to use the image available in the this software-orchestration blueprint.

File: eclipse-ankaios/README.md
Comment (inf17101) [reply eclipse-ankaios/README.md]: @eriksven: It is renamed now and we made some other improvements inside the documentation and Dockerfile. Pls have a look again.

**Comment by inf17101** [comment]
> I merged #15 and triggered the workflow for building the image. The image is now available in https://github.com/eclipse-sdv-blueprints/software-orchestration/pkgs/container/software-orchestration%2Feclipse-chariott%2Fservice-discovery . Could you adapt this PR to use the image available in the this software-orchestration blueprint.

@eriksven and @ladatz: We have set the new image path and have started it. It works.

**Comment by eriksven** [comment]
That is great news. I guess it then makes sense to wait for #17 to be merged. We could then run the new workflows for the custom containers mqtt-broker-agemo-config:0.1.2, eclipse-agemo/pub-sub-service:0.1.2, and eclipse-ibeji/invehicle-digital-twin:0.1.1 and adapt this PR accordingly. 
@ladatz do you intend to have a workflow for building the eclipse-ibeji/invehicle-digital-twin:0.1.0 container too?

**Comment by ladatz** [comment]
> That is great news. I guess it then makes sense to wait for #17 to be merged. We could then run the new workflows for the custom containers mqtt-broker-agemo-config:0.1.2, eclipse-agemo/pub-sub-service:0.1.2, and eclipse-ibeji/invehicle-digital-twin:0.1.1 and adapt this PR accordingly. @ladatz do you intend to have a workflow for building the eclipse-ibeji/invehicle-digital-twin:0.1.0 container too?

Yes there will be a workflow for the final image (freyja), likely PR out today. Then we will have the full in-vehicle stack. The sample applications will also be added

**Comment by inf17101** [comment]
A new release of Ankaios is published v0.3.0. We will add a new container image, update the image config in this PR and also adapt the startup state config to support the new features. Then we can get rid of the additional helper workload and start the stuff directly. I have already tested the inter-workload dependencies features last month during development (a separate branch exists on this repo here) and it has worked, but I wanted to check again that all works with the latest changes together in this branch. We will push the changes next week.

File: eclipse-ankaios/config/startupState.yaml
Comment (eriksven) [inline eclipse-ankaios/config/startupState.yaml:36]: ghcr.io/eclipse-sdv-blueprints/software-orchestration/eclipse-agemo/pub-sub-service:0.1.2

File: eclipse-ankaios/config/startupState.yaml
Comment (eriksven) [inline eclipse-ankaios/config/startupState.yaml:48]: ghcr.io/eclipse-sdv-blueprints/software-orchestration/eclipse-ibeji/invehicle-digital-twin:0.1.1

File: eclipse-ankaios/config/startupState.yaml
Comment (eriksven) [inline eclipse-ankaios/config/startupState.yaml:60]: ghcr.io/eclipse-sdv-blueprints/software-orchestration/eclipse-freyja/local-with-ibeji:0.1.0

File: eclipse-ankaios/config/startupState.yaml
Comment (eriksven) [inline eclipse-ankaios/config/startupState.yaml:10]: ghcr.io/eclipse-sdv-blueprints/software-orchestration/eclipse-chariott/service-discovery:0.2.1

File: eclipse-ankaios/config/startupState.yaml
Comment (eriksven) [inline eclipse-ankaios/config/startupState.yaml:22]: ghcr.io/eclipse-sdv-blueprints/software-orchestration/eclipse-mosquitto/mqtt-broker-agemo-config:0.1.2

(most likely)

**Comment by inf17101** [comment]
> @inf17101 we waited with the merge of this PR because the referenced container images were not available in this repository yet. With the merge of #20 this has now been resolves. Do you want to update your PR accordingly?
> 
> Are there other blockers that prevent us from merging this PR then? I am asksing because in the first comment it says that the PR should not be merged.

@eriksven, @ladatz: Thank you for your updates. I think we can merge the PR with the initial config when all services are working. I tried on MacOS quickly because it already after work, and I can see that with the images located at "ghcr.io/ladatz" as well as the new images built inside the repository "ghcr.io/eclipse-sdv-blueprints/software-orchestration" two services are not coming up. So they failed. 

Here a screenshot of the ank-cli:
Bildschirmfoto 2024-05-17 um 20 25 35

Here are the details of the podman logs:
Bildschirmfoto 2024-05-17 um 20 27 23

Maybe we can fix this and then I think we can merge this PR.

@eriksven: I have seen that at the very first conversations I wait for feedback of you. I have fixed some findings previously, but conversation is still open. Maybe you could have a look again? That would be nice.

**Comment by ladatz** [comment]
> > @inf17101 we waited with the merge of this PR because the referenced container images were not available in this repository yet. With the merge of #20 this has now been resolves. Do you want to update your PR accordingly?
> > Are there other blockers that prevent us from merging this PR then? I am asksing because in the first comment it says that the PR should not be merged.
> 
> @eriksven, @ladatz: Thank you for your updates. I think we can merge the PR with the initial config when all services are working. I tried on my private Mac quickly because it is already after work. I can see that with the images located at "ghcr.io/ladatz" as well as the new images built inside the repository "ghcr.io/eclipse-sdv-blueprints/software-orchestration" two services are not coming up. So they failed.
> 
> Here a screenshot of the ank-cli: Bildschirmfoto 2024-05-17 um 20 25 35
> 
> Here are the details of the podman logs: Bildschirmfoto 2024-05-17 um 20 27 23
> 
> Maybe we can fix this and then I think we can merge this PR.
> 
> @eriksven: I have seen that at the very first conversations I wait for feedback of you. I have fixed some findings previously, but conversation is still open. Maybe you could have a look again? That would be nice.

@inf17101 Thanks for the update, I will pull your branch and take a look at the failing images

**Comment by ladatz** [comment]
> > > @inf17101 we waited with the merge of this PR because the referenced container images were not available in this repository yet. With the merge of #20 this has now been resolves. Do you want to update your PR accordingly?
> > > Are there other blockers that prevent us from merging this PR then? I am asksing because in the first comment it says that the PR should not be merged.
> > 
> > 
> > @eriksven, @ladatz: Thank you for your updates. I think we can merge the PR with the initial config when all services are working. I tried on my private Mac quickly because it is already after work. I can see that with the images located at "ghcr.io/ladatz" as well as the new images built inside the repository "ghcr.io/eclipse-sdv-blueprints/software-orchestration" two services are not coming up. So they failed.
> > Here a screenshot of the ank-cli: Bildschirmfoto 2024-05-17 um 20 25 35
> > Here are the details of the podman logs: Bildschirmfoto 2024-05-17 um 20 27 23
> > Maybe we can fix this and then I think we can merge this PR.
> > @eriksven: I have seen that at the very first conversations I wait for feedback of you. I have fixed some findings previously, but conversation is still open. Maybe you could have a look again? That would be nice.
> 
> @inf17101 Thanks for the update, I will pull your branch and take a look at the failing images

@inf17101 I pulled the changes and ran it on a Mac with M1 chip with the eclipse-sdv-blueprints images. All of the services started up. Please make sure you remove any old images that may be dangling, and let me know if you still have issues. 
[services.pdf](https://github.com/eclipse-sdv-blueprints/software-orchestration/files/15358381/services.pdf)

**Comment by inf17101** [comment]
> > > > @inf17101 we waited with the merge of this PR because the referenced container images were not available in this repository yet. With the merge of #20 this has now been resolves. Do you want to update your PR accordingly?
> > > > Are there other blockers that prevent us from merging this PR then? I am asksing because in the first comment it says that the PR should not be merged.
> > > 
> > > 
> > > @eriksven, @ladatz: Thank you for your updates. I think we can merge the PR with the initial config when all services are working. I tried on my private Mac quickly because it is already after work. I can see that with the images located at "ghcr.io/ladatz" as well as the new images built inside the repository "ghcr.io/eclipse-sdv-blueprints/software-orchestration" two services are not coming up. So they failed.
> > > Here a screenshot of the ank-cli: Bildschirmfoto 2024-05-17 um 20 25 35
> > > Here are the details of the podman logs: Bildschirmfoto 2024-05-17 um 20 27 23
> > > Maybe we can fix this and then I think we can merge this PR.
> > > @eriksven: I have seen that at the very first conversations I wait for feedback of you. I have fixed some findings previously, but conversation is still open. Maybe you could have a look again? That would be nice.
> > 
> > 
> > @inf17101 Thanks for the update, I will pull your branch and take a look at the failing images
> 
> @inf17101 I pulled the changes and ran it on a Mac with M1 chip with the eclipse-sdv-blueprints images. All of the services started up. Please make sure you remove any old images that may be dangling, and let me know if you still have issues. [services.pdf](https://github.com/eclipse-sdv-blueprints/software-orchestration/files/15358381/services.pdf)

Hi @ladatz: I have checked out the repo first times on my Mac M1 on friday (so no dangling images there at all). I have also tried today again and the same two images are failing. I have also tested on my Linux Manjaro and there also Freyja image fails (but only 1 image not 2 like on the Mac). I am still checking what's the issue.

File: eclipse-ankaios/README.md
Comment (eriksven) [reply eclipse-ankaios/README.md]: Looks good to me, thank you!

**Comment by eriksven** [comment]
@inf17101 I looked at the current state of the PR and it looks good to me. There was one question to which I responded but it is easy to get lost in the different threads. So let me know if I missed another conversation or question.

File: eclipse-ankaios/.devcontainer/devcontainer.json
Comment (inf17101) [reply eclipse-ankaios/.devcontainer/devcontainer.json]: I have synced now and merged main into this branch. Unfortunately, I still have the same issues with the failing services.

Comment (inf17101) [reply eclipse-ankaios/.devcontainer/devcontainer.json]: @ladatz: Even after merging current main into this branch, I have the same issues with freyja and ibieji on Mac M1 and Freyja on Linux. I have had a few minutes time and have treid out the following:
- Starting all services besides the freyja image with `run_blueprint.sh` inside the devcontainer
- Run the freyja image manually after all service started with the script above are running and executed the entrypoint of the container image manually

The first times it fails again with the already shared panic and the error about the missing URI in the GRPCServiceDiscoveryAdapter config. Then I just hit Ctrl+C and just executed the same command to start freyja again and then it works and receives the weight properties. I do not know why it works all the time after the second start but here is a screenshot of the logs:

image

Comment (inf17101) [reply eclipse-ankaios/.devcontainer/devcontainer.json]: @ladatz, @eriksven: Today I tested again on the Mac M1 and it seems to work now. I repeated the tests several times. Did you change anything on the pushed images? I think we can merge this PR then and we are keeping track if it is sporadically failing or not.

Comment (ladatz) [reply eclipse-ankaios/.devcontainer/devcontainer.json]: Nothing has changed on the images after they were published in March and May. If we observe any failures in the future we can open issues/ discussions to resolve them. 
image

File: eclipse-ankaios/config/startupState.yaml
Comment (ladatz) [reply eclipse-ankaios/config/startupState.yaml]: Resolving since the images from the blueprint container registry are used now

**Comment by inf17101** [comment]
> @inf17101 I looked at the current state of the PR and it looks good to me. There was one question to which I responded but it is easy to get lost in the different threads. So let me know if I missed another conversation or question.

Ok thank you. @ladatz @eriksven, I will do one last change, upgrading to the latest Ankaios version v0.3.1 and test if all is working again. Then we can merge.

**Comment by ladatz** [comment]
> > @inf17101 I looked at the current state of the PR and it looks good to me. There was one question to which I responded but it is easy to get lost in the different threads. So let me know if I missed another conversation or question.
> 
> Ok thank you. @ladatz @eriksven, I will do one last change, upgrading to the latest Ankaios version v0.3.1 and test if all is working again. Then we can merge. By the way shall the smart trailer app also work in the initial contribution? Or just the services up and running?
> 
> I get the following output when testing the smart trailer script:
> 
> image
> 
> Checked with a little telnet afterwards and Ibeji is reachable from within the script. The entity does not exist!
> 
> However when executing the ank cli commands of the smart trailer script manually, then the smart trailer app and provider seems to log some correct output and are up and running:
> 
> image
> 
> Also the freyja outputs valid trailer weight data.
> 
> I think it looks like that the "start trailer" script fails at the beginning when checking the smart trailer `dtmi:sdv:Trailer:IsTrailerConnected;1` property. And that fails somehow because the entity does not exist.

As discussed offline, this is because part of the demo is to run the command to start the trailer_connected_provider. I will merge this PR (it has been tested on both Linux and Mac). I will also propose a follow-up change to add this command to the README.md

**Review by ladatz — APPROVED** [review]

**Pull request merged**
